crates/api/src/relay_data/api.rs

```diff
@@ -131,15 +141,48 @@ impl<DB: DatabaseService + 'static> DataApi<DB> {
         }
     }
 
-    pub async fn constraints() {
-        unimplemented!()
+    /// Implements this API: <https://chainbound.github.io/bolt-docs/api/relay#constraints>
+    pub async fn constraints(
+        Extension(data_api): Extension<Arc<DataApi<A, DB>>>,
+        Query(params): Query<usize>,
+    ) -> Result<impl IntoResponse, DataApiError> {
+        // TODO: Return err if the slot is outside expiry bounds i.e. before an epoch
+        // For this we need a beacon state tracker
+        let slot = params as u64;
+        
+        match data_api.auctioneer.get_constraints(slot).await {
+            Ok(Some(constraints_with_proof_data)) => {
+                let constraints = constraints_with_proof_data
+                    .into_iter()
+                    .map(|data| data.message)
+                    .collect::<Vec<ConstraintsMessage>>();
+        
+                Ok(Json(constraints))
+            }
+            Ok(None) => {
+                Ok(Json(vec![])) // Return an empty vector if no constraints are found
+            }
+            Err(err) => {
+                warn!(error=%err, "Failed to get constraints");
+                Err(DataApiError::ConstraintsError(slot))
+            }
+        }
     }
 
-    pub async fn constraints_stream() {
-        unimplemented!()
-    }
+    /// Implements this API: <https://chainbound.github.io/bolt-docs/api/relay#constraints-stream>
+    pub async fn constraints_stream(
+        Extension(data_api): Extension<Arc<DataApi<A, DB>>>,
+    ) -> Sse<impl Stream<Item = Result<Event, DataApiError>>> {
+        let constraints_rx = data_api.constraints_rx.read().await;
 
-    // pub async fn blocks_with_proofs() {
-    //     unimplemented!()
-    // }
+        let filtered = constraints_rx.map(|constraint| match serde_json::to_string(&constraint) {
+            Ok(json) => Ok(Event::default().data(json).event("constraint").retry(Duration::from_millis(50))),
+            Err(err) => {
+                warn!(error=%err, "Failed to serialize constraint");
+                Err(DataApiError::InternalServerError)
+            }
+        });
+
+        Sse::new(filtered).keep_alive(KeepAlive::default())
+    }
 }
```

Comment:
How does this SSE stay open? I don't fully understand how long we are keeping the read() lock for when opening a stream. What happens if we want to write in the constraints while also streaming contents?

Reply:
Thanks for pointing out, resolved